[Translation] Added intl message formatter.

Nyholm opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | replaces #20007
| License       | MIT
| Doc PR        | 

This PR will replace #20007 and continue the work from the original author. 

I've have tried to address all comments made in the original PR.

aitboudad committed:
[Translation] Added intl message formatter.

Nyholm committed:
Add support for multiple formatters

Nyholm committed:
Allow config for different domain specific formatters

frankdejonge commented on src/Symfony/Component/Translation/Formatter/IntlMessageFormatter.php:
```diff
@@ -0,0 +1,45 @@
+     */
+    public function format($message, $locale, array $parameters = array())
+    {
+        $formatter = new \MessageFormatter($locale, $message);
```
When the parsing of this message fails you get a very non-descriptive exception message (`Constructor failed`). Might be good for DX to rethrow a more descriptive one.

frankdejonge commented on src/Symfony/Component/Translation/Translator.php:
```diff
@@ -434,4 +444,17 @@ private function getConfigCacheFactory(): ConfigCacheFactoryInterface
+     */
+    private function getFormatter(string $domain)
+    {
+        if (isset($this->formatters[$domain])) {
```
this could be reduced to:


```php
return $this->formatters[$domain] ?? $this->formatters['_default'];
```

Koc replied:
And we can declare return type also

Koc commented on src/Symfony/Component/Translation/Tests/Formatter/IntlMessageFormatterTest.php:
```diff
@@ -0,0 +1,91 @@
+
+    public function testFormatWithNamedArguments()
+    {
+        if (PHP_VERSION_ID < 50500 || version_compare(INTL_ICU_VERSION, '4.8', '<')) {
```
Checking for php version is redundant

https://github.com/symfony/symfony/blob/f557f943babf2893442ceb240521ef1bde785c50/composer.json#L19

Nyholm committed:
Make it a warning

Nyholm committed:
use error

nicolas-grekas commented on src/Symfony/Component/Translation/Translator.php:
```diff
@@ -132,6 +132,15 @@ public function addResource($format, $resource, $locale, $domain = null)
 
+    /**
+     * @param string                    $domain
+     * @param MessageFormatterInterface $formatter
```
to be removed :)

Nyholm replied:
Thank you. I added `: void` as return type.

nicolas-grekas commented on src/Symfony/Component/Translation/Translator.php:
```diff
@@ -434,4 +444,14 @@ private function getConfigCacheFactory(): ConfigCacheFactoryInterface
+    /**
+     * @param string $domain
+     *
+     * @return MessageFormatterInterface
```
to be removed (and a real  return type be used instead)

Nyholm replied:
Thank you

Nyholm commented:
Anything I can do to help review this?

artursvonda commented:
Love the change and allowing to get rid of special case of transChoice (over time).

The original PR mentioned something about storing format with the messages and I actually tend to agree with it. The format IS information that's directly coupled with actual messages (basically, a meta information about contents of the file). This becomes even more important if translations are coming from external source and are not part of source files. There already have been talks about allowing different caching mechanisms and sources (like database) so coupling domain to format in config would not allow for changes in message format without deployment (and would make syncing changes hard anyway).
Catalogues already support metadata so we can add formatter as additional key there and maybe think about making formatter metadata part of actual message which would allow even more granular transition of messages from one format to another.
Architecture wise this would mean a bigger change long term with messages probably being objects (or structs) containing metadata but short term we can just use similar strategy as in this PR but configure domain => formatter relation automatically based on metadata in translation files.

nicolas-grekas commented:
Added to project "Contracts" after comment https://github.com/symfony/symfony/pull/28210#discussion_r212756794

Koc commented on src/Symfony/Component/Translation/Formatter/IntlMessageFormatter.php:
```diff
@@ -0,0 +1,49 @@
+     */
+    public function choiceFormat($message, $number, $locale, array $parameters = array())
+    {
+        return $this->format($message, $locale, $parameters);
```
Is this kind of implementation right? Argument `number` is unused.

jvasseur replied:
I think we shouldn't implement the `ChoiceMessageFormatterInterface` instead, it will disallow using the `transChoice` method which is right since we should only use the `trans` method for intl-formated message.

Nyholm commented on src/Symfony/Component/Translation/Formatter/IntlMessageFormatter.php:
```diff
@@ -0,0 +1,49 @@
+        } catch (\Throwable $e) {
+            throw new \InvalidArgumentException('Invalid message format.', $e);
+        }
+        if (null === $formatter) {
```
It will never be null here

nicolas-grekas commented:
I like it, but I would make the new formatter accessible only through `trans()` and not through `transChoice()` which should support only the Symfony syntax.
This would pave the way to deprecating transChoice at a later point in time, following [this comment](https://github.com/symfony/symfony/pull/28210#discussion_r212756794) from @Koc.
I'd suggest to wait for https://github.com/symfony/symfony/pull/28210 to be merged before rebasing this PR and resolving this very comment.

nicolas-grekas commented:
Some description of the format here:
https://messageformat.github.io/messageformat/page-guide

Nyholm committed:
Fixes according to feedback

Nyholm committed:
Use FallbackFormatter instead of support for multiple formatters

Nyholm commented:
I've updated this PR. 
I did some major changes. I removed support for multiple formatters. Instead, I added a `FallbackFormatter`. 

With the fallback formatter we can support both "Symfony style plural" and intl messages at the same time. This will make sure we have a zero config solution for adding support for this new feature while all the old formats still works as expected. 

Possible update: I could make the `FallbackFormatter` less generic by using the following constructor signature: `FallbackFormatter::__construct(IntlMessageFormatter $first, MessageFormatter $second)`. The logic will be simpler if I do this change.

fabpot commented on src/Symfony/Component/Translation/Tests/Formatter/IntlMessageFormatterTest.php:
```diff
@@ -0,0 +1,98 @@
+        if (!\extension_loaded('intl')) {
+            $this->markTestSkipped(
+              'The Intl extension is not available.'
+            );
```
should be on one line

fabpot commented on src/Symfony/Component/Translation/Tests/Formatter/IntlMessageFormatterTest.php:
```diff
@@ -0,0 +1,98 @@
+
+class IntlMessageFormatterTest extends \PHPUnit\Framework\TestCase
+{
+    public function setUp()
```
should be protected

fabpot commented on src/Symfony/Component/Translation/Tests/Formatter/IntlMessageFormatterTest.php:
```diff
@@ -0,0 +1,98 @@
+        );
+    }
+
+    private function getMessageFormatter()
```
I think this can be inlined instead.

fabpot commented on src/Symfony/Component/Translation/CHANGELOG.md:
```diff
@@ -7,6 +7,7 @@ CHANGELOG
  * Started using ICU parent locales as fallback locales.
  * deprecated `TranslatorInterface` in favor of `Symfony\Contracts\Translation\TranslatorInterface`
  * deprecated `MessageSelector`, `Interval` and `PluralizationRules`; use `IdentityTranslator` instead
+ * Added `IntlMessageFormatter` and`FallbackMessageFormatter`
```
missing space after `and`.

fabpot commented on src/Symfony/Component/Translation/Formatter/FallbackFormatter.php:
```diff
@@ -0,0 +1,78 @@
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
```
AFAIK, we are not using this anywhere else, this should be removed.

fabpot commented on src/Symfony/Component/Translation/Formatter/FallbackFormatter.php:
```diff
@@ -0,0 +1,78 @@
+            return $this->secondFormatter->choiceFormat($message, $number, $locale, $parameters);
+        }
+
+        throw new LogicException(sprintf('The no formatter support plural translations.'));
```
The message is not so clear to me :) I suppose you meant `No formatters support plural translations.`

fabpot commented on src/Symfony/Component/Translation/Formatter/IntlMessageFormatter.php:
```diff
@@ -0,0 +1,40 @@
+        try {
+            $formatter = new \MessageFormatter($locale, $message);
+        } catch (\Throwable $e) {
+            throw new InvalidArgumentException(sprintf('Invalid message format. Reason: %s (error #%d)', intl_get_error_message(), intl_get_error_code()), 0, $e);
```
As an exception should end with a dot, I suggest to use `Invalid message format (%s, error #%d).` Same below

Nyholm committed:
Updates according to feedback

Nyholm commented:
Thank you for the review, Ive updated the PR

Nyholm committed:
Only use the default translator if intl extension is loaded

nicolas-grekas commented on src/Symfony/Component/Translation/Formatter/FallbackFormatter.php:
```diff
@@ -0,0 +1,76 @@
+    {
+        try {
+            $result = $this->firstFormatter->format($message, $locale, $parameters);
+        } catch (\Throwable $e) {
```
Can we be more specific here? We should be as much as possible IMHO (same below.)

Nyholm replied:
Do we want to be more specific?
I know the IntlMesaageFormatter only throws InvalidArgument, but isn’t it a good idea to catch everything to give the second formatter a change to run?

Nyholm committed:
Be more specific with what exception we catch

fabpot commented on src/Symfony/Component/Translation/Tests/Formatter/FallbackFormatterTest.php:
```diff
@@ -17,6 +17,7 @@
 use Symfony\Component\Translation\Formatter\FallbackFormatter;
 use Symfony\Component\Translation\Formatter\MessageFormatterInterface;
 
+
```
should be removed

fabpot commented on src/Symfony/Component/Translation/Tests/Formatter/FallbackFormatterTest.php:
```diff
@@ -0,0 +1,214 @@
+use Symfony\Component\Translation\Formatter\FallbackFormatter;
+use Symfony\Component\Translation\Formatter\MessageFormatterInterface;
+
+
```
should be reverted

fabpot reviewed (approved):
With one small comment. In a followup PR, we should deprecate the old way, deprecate `transChoice` and probably try to write a script that converts our proprietary format to the intl one. WDYT?

nicolas-grekas commented:
> In a followup PR, we should deprecate the old way, deprecate transChoice and probably try to write a script that converts our proprietary format to the intl one

:100: for both!

Nyholm committed:
cs

Nyholm commented:
Thank you. PR updated. 

I agree to deprecate transChoice

fabpot commented:
Thank you @Nyholm.
